### PR TITLE
v5.0.x: Safety valve for callbacks

### PR DESCRIPTION
**Pull request by awlauria**
When an mca_foo.so gets unloaded any callbacks it registered become invalid and can
segv if called.

This happened in an oshemem testcase that did call shmem_init() and shmem_finalize().
But for whatever reason oshmem unloads the MCAs before they call MPI_Finalize() which
they have in a conditional so they don't necessarily always call MPI_Finalize().
Anyway the failing path goes from oshmem_shmem_finalize() through _shmem_finalize()
which unloads the MCAs (mca_base_framework_close which dlcloses them) and then after
that it will call MPI_Finalize() which will deregister the callback

That could be considered a bug, maybe the reference counters should be set up so the
MCA's .fini call is always called before dlclose() and make sure all such .fini calls
deregister the MCA's callbacks.

But this checkin takes the conservative approach of assuming that bugs like that are
always a possibility and inserts a safety valve around callback usage, so when an MCA
gets unloaded it deregisters any callbacks rather than only having them deregistered
if we go through a proper sequence of finalize calls.

Signed-off-by: Austen Lauria <awlauria@us.ibm.com>
(cherry picked from commit 86b3d100d955a17dce17bdbb1f7d805653c91c6a)

bot:notacherrypick

### CODE REVIEW

**Comment by awlauria** [comment]
bot:aws:retest

**Comment by gpaulsen** [comment]
@hppritcha Will this continue to work when Sessions is merged in?
I.e. When using sessions and doing MPI_Session_init in two different threads, is there a mutex somewhere to order the threads through the various mca component initialization?  Or does it just go through mca initialization once?
If not, we may need a lock around the static variable used in this PR.

**Review by hppritcha — COMMENTED** [review]
I think @gpaulsen has a point.  this safety_value variable should best be protected by a static mutex as a precaution.

**Comment by markalle** [comment]
I don't think mutexes would be needed for the safety valve variable.  We could probably even go the opposite direction and remove the "use_safetey_valve" variable entirely.  It's only there because I didn't like the idea of calling "unregister_release" if we hadn't ever called "register_release".  But it's not doing any reference counting to make sure the number of registers and unregisters is equal.  It's just saying that as a library-level destructor, it's going to make sure there's a final unregister before the callback function pointers become invalid regardless of whether the rest of OMPI has its reference counters all working properly to match up all the register/unregisters or not.

**Comment by markalle** [comment]
Austin made a nice cleanup over in 9853, removing the unnecessary tracking variable. So I think this plus 9853 is the way to go

**Comment by gpaulsen** [comment]
@markalle this was put into Draft mode, because the new static variable might cause issues with MPI Sessions (and MPI 4.0 feature).

Could you please add a thread lock around it’s use? (can look at ompi_mpi_instance_init in ompi/instance/instance.c for example).

Then PR this new commit to main in a new PR and also push this commit to the end of https://github.com/open-mpi/ompi/pull/9748 please. :slightly_smiling_face:
Then once that passes CI and is merged to main we can remove the draft label from the v5.0 PR #9748 and get that in for v5.0.0.

**Comment by gpaulsen** [comment]
This is v5.0.x version of https://github.com/open-mpi/ompi/pull/10728 on `main`.

**Review by gpaulsen — APPROVED** [review]
I approve, assuming that `__attribute__((destructor));` is available across all compilers.

**Comment by markalle** [comment]
Even though __attribute__ support I think is widespread, I think you've got a point that the more correct solution is to use ```__opal_attribute_destructor__``` which would #ifdef it away for strange compilers without attrbitues, and I think that's as good a solution as any.  Such a compiler would then lack the safety valve functionality in its OMPI build, but I think that's just what you get if the compiler doesn't have the support

**Comment by markalle** [comment]
I made a different PR
https://github.com/open-mpi/ompi/pull/10733
that should be the same as this, but is now just made from two cherry-picks out of main

**Pull request merged**
